### PR TITLE
Migrate git/branches.go to git.Runner

### PR DESCRIPTION
**Pull request by kevgo**
This migrates all legacy Git methods to the new world. Progress!

### CODE REVIEW

File: src/cmd/shared.go [hunk 1]
```diff
@@ -16,6 +16,7 @@ var (
 	debugFlag,
 	dryRunFlag,
 	globalFlag bool
+	repoCache *git.ProdRepo
```
Comment (charlierudolph) [inline src/cmd/shared.go:19]: Thoughts on making this a const called `repo` that is initialized? Not sure if the constructor has side effects we don't want to run at other times

Comment (charlierudolph) [reply src/cmd/shared.go:19]: Though I guess that may introduce confusion as repo is a local variable in many functions in this package. Maybe call it `prodRepo` then?

File: src/git/prod_repo.go [hunk 2]
```diff
@@ -26,3 +26,21 @@ func NewProdRepo() *ProdRepo {
...
+		hasChildBranch, err := r.Silent.HasLocalOrRemoteBranch(child)
+		if err != nil {
+			return err
+		}
+		hasParentBranch, err := r.Silent.HasLocalOrRemoteBranch(parent)
+		if err != nil {
+			return err
+		}
+		if !hasChildBranch || !hasParentBranch {
+			r.Configuration.DeleteParentBranch(child)
```
Comment (charlierudolph) [inline src/git/prod_repo.go:42]: Interesting. Just thinking on this more how Configuration has its own shell. Any thoughts if there are other "libraries" like within a gitRepo we would want to break out to function similar to Configuration? Would probably need to be things that always run silently.

Conversely, thoughts on removing Configuration as a special case?

File: src/git/runner.go [hunk 3]
```diff
@@ -682,6 +683,84 @@ func (r *Runner) LocalAndRemoteBranches() ([]string, error) {
...
+	for _, line := range res.OutputLines() {
+		line = strings.Trim(line, "* ")
+		line = strings.TrimSpace(line)
+		result = append(result, line)
+	}
+	return
+}
+
+// LocalBranchesMainFirst provides the names of all local branches in this repo.
+func (r *Runner) LocalBranchesMainFirst() (result []string, err error) {
```
Comment (charlierudolph) [inline src/git/runner.go:702]: I'm confused about the difference between this and `LocalBranchesWithMainBranchFirst`

File: src/cmd/shared.go (hunk 1 above)
Comment (kevgo) [reply src/cmd/shared.go:19]: Great idea! All commands need this variable anyways. I went with `prodRepo`.

File: src/git/prod_repo.go (hunk 2 above)
Comment (kevgo) [reply src/git/prod_repo.go:42]: Yes, the new architecture consists of packages that are orthogonal to each other and fully independent of each other, i.e. no dependencies between them. In the ideal case that I try to reach, the `git` package defines a `git.Shell` interface that describes the shell methods that the `git` package needs to execute. It is then given an instance of the `command.SilentShell` struct that implements this interface. This means there are no dependencies between the `git` and the `command` packages but they compose well and integrate each other deeply. 

There might be additional aspects that could be extracted into such independent, orthogonal, composable packages. My current list is [here](https://github.com/git-town/git-town/blob/master/documentation/development/architecture.md). I think you are right that the configuration is independent from running Git commands. The fact that we store configuration in Git is just an implementation detail. There should probably be a `config` package.

It's interesting that the Go syntax can be learned in a week, but it took us (or at least me) years to understand how to use it the right way. 😃 It's a bit like the game Go in this sense. Or maybe Cyvasse!

File: src/git/runner.go (hunk 3 above)
Comment (kevgo) [reply src/git/runner.go:702]: Rightfully so, they are the same! Thanks for catching this.

Comment (kevgo) [reply src/git/runner.go:702]: You are - as always - refreshingly meticulous!

**Comment by kevgo** [comment]
Really great to hear from you @charlierudolph! All comments addressed, I'm going to merge this in as usual to maintain momentum. Happy to address further ideas and comments separately. This is far from done!

**Pull request merged**